[cmake/cpack] Debian packages generation

hudokkow opened the pull request:
This is the work to generate debian packages with our shiny CMake/CPack system.

It currently generates deb packages following stable, unstable or nightly release conventions for name/version:
kodi-bin_16.1~git20160425.1001-final-0xenial_amd64.deb   / 2:16.1~git20160425.1001-final-0xenial
kodi-bin_17.0~git20160703.1014-alpha2-0xenial_amd64.deb  / 2:17.0~git20160703.1014-alpha2-0xenial
kodi-bin_17.0~git20160714.0200-b0970fe-0xenial_amd64.deb / 2:17.0~git20160714.0200-b0970fe-0xenial
##### Packages currently generated

kodi
kodi-addon-dev
kodi-audio-dev
kodi-bin
kodi-eventclients-common
kodi-eventclients-dev
kodi-eventclients-ps3
~~kodi-eventclients-j2me~~
kodi-eventclients-wiiremote
kodi-eventclients-xbmc-send
kodi-inputstream-dev
kodi-pvr-dev
kodi-screensaver-dev
kodi-tools-texturepacker
kodi-visualization-dev
~~xbmc~~ removed
### Design decisions
- ~~CPackConfig.cmake file contains variables and logic common to all generators (DEB/RPM/NSIS/etc.)~~
- CPackConfigDEB.cmake file contains variables and logic specific to DEB packaging system
- scripts/linux/Install.cmake handles install targets logic separated by components. It calls (includes) CPackConfigDEB.cmake for package generation. ~~In turn CPackConfigDEB.cmake is the only file calling CPackConfig.cmake~~
- Variables and logic are enclosed in if statements. Should provide an easy way to manipulate packages naming/version/etc. using command-line only.
- CMake/CPack >=3.6.0 is required for proper package naming generation because it introduced <code> CPACK_DEBIAN_component_name_FILE_NAME</code>. It will work with lower versions but resulting package names are meh.
### Todo
- [x] Add documentation to packages (copyright, license, etc.)
- [x] Generate debian changelog on-the-fly and add it to packages (ping @wsnipex for ideas)
- [x] ~~Generate xbmc -> kodi symlinks and add them to packages~~ symlinks removed. They mess cpack DEB generation and xbmc to kodi transition as lasted long enough.
- [x] ~~Generate kodi-eventclients-j2me package (I saw that j2me-remote needs ant and immediately lost motivation. :wink: I'm sure I'll find it again after I'm done with all the TODOs)~~ Support for this client was removed in https://github.com/xbmc/xbmc/pull/10049
- [x] Generate kodi-eventclients-ps3
- [x] Generate kodi-eventclients-wiiremote
- [x] Generate xbmc (dummy package)
- [x] Generate source tar.gz
- [x] Cleanup files
- [x] Add documentation to README

As always opinions are welcome.

cc/ @wsnipex, @fetzerch, @notspiff and anyone else interested

hudokkow committed:
[cmake] Add functions required for CPack DEB

hudokkow committed:
[cmake] Refactor linux Install.cmake by components

hudokkow committed:
[cpack] Add CPackConfigDEB.cmake with specific DEB variables and logic

akva2 commented on project/cmake/scripts/linux/Install.cmake:
```diff
@@ -38,98 +38,273 @@ configure_file(${CORE_SOURCE_DIR}/tools/Linux/kodi-standalone.sh.in
 configure_file(${PROJECT_SOURCE_DIR}/KodiConfig.cmake.in
                ${CORE_BUILD_DIR}/scripts/${APP_NAME}Config.cmake @ONLY)
+configure_file(${PROJECT_SOURCE_DIR}/XBMCConfig.cmake.in
+               ${CORE_BUILD_DIR}/scripts/XBMCConfig.cmake @ONLY)
```
This file was kept for backwards compatibility. Somebody then went ahead and renamed it so it serves no purpose. Nuke or bring back old naming (only one with much legacy is me and i have updated all my stuff so i vote nuke.

notspiff commented on project/cmake/CPackConfig.cmake:
```diff
@@ -0,0 +1,59 @@
+  set(CPACK_PACKAGE_VERSION_PATCH "0")
+endif()
+if(NOT CPACK_PACKAGE_DESCRIPTION)
+  set(CPACK_PACKAGE_DESCRIPTION "\
```
resource file please.

notspiff commented on project/cmake/CPackConfigDEB.cmake:
```diff
@@ -0,0 +1,267 @@
+set(CPACK_COMPONENT_KODI_VISUALIZATION_DEV_DESCRIPTION "\
+Kodi Media Center (visualization add-on dev package)\n\
+${CPACK_PACKAGE_DESCRIPTION}\n \
+This is the Development package for Kodi visualization add-ons.")
```
i'd like to see all of this in a resource control-like files. much lower barrier to changing things (no need to grok cmake syntax).

hudokkow replied:
Uhm... are you talking about using <code>CPACK_DEBIAN_component_name_PACKAGE_CONTROL_EXTRA</code> with per component control file?

If I understand it correctly, the control file is an all or nothing approach, i.e. all fields must be populated there.

That brings in another step:
configure control_component_name.in files to generate a control file per component with 
Version: ${PACKAGE_VERSION}, Architecture: ${CPACK_SYSTEM_NAME} and perhaps also Maintainer: ${PACKAGE_MAINTAINER} (or alternatively hardcode maintainer) during configuration.

Is that what you mean?

notspiff replied:
I mean store the metadata in text files, not literary a control file a
such. Costs some more cmake code now, but it means separation of code and
data and in the long term that is always a good precaution to avoid stupid
mistakes where code is changed by accident. The same reasoning is behind a
lot of the buildsystem, paths separated from code etc.
16. jun. 2016 15:56 skrev "Cristiano A. Silva" <notifications@github.com>:

> In project/cmake/CPackConfigDEB.cmake
> <https://github.com/xbmc/xbmc/pull/9987#discussion_r67348291>:
>
> > +set(CPACK_COMPONENT_KODI_SCREENSAVER_DEV_DESCRIPTION "\
> > +Kodi Media Center (screensaver add-on dev package)\n\
> > +${CPACK_PACKAGE_DESCRIPTION}\n \
> > +This is the Development package for Kodi screensaver add-ons.")
> > +
> > +# kodi-visualization-dev package
> > +set(CPACK_DEBIAN_KODI_VISUALIZATION_DEV_FILE_NAME ${CPACK_PACKAGE_NAME}-visualization-dev_${PACKAGE_NAME_VERSION}_all.deb)
> > +set(CPACK_DEBIAN_KODI_VISUALIZATION_DEV_PACKAGE_NAME ${CPACK_PACKAGE_NAME}-visualization-dev)
> > +set(CPACK_DEBIAN_KODI_VISUALIZATION_DEV_PACKAGE_DEPENDS "kodi-addon-dev")
> > +set(CPACK_DEBIAN_KODI_VISUALIZATION_DEV_PACKAGE_SECTION libdevel)
> > +set(CPACK_DEBIAN_KODI_VISUALIZATION_DEV_PACKAGE_PRIORITY optional)
> > +set(CPACK_DEBIAN_KODI_VISUALIZATION_DEV_PACKAGE_PROVIDES "xbmc-visualization-dev")
> > +set(CPACK_COMPONENT_KODI_VISUALIZATION_DEV_DESCRIPTION "\
> > +Kodi Media Center (visualization add-on dev package)\n\
> > +${CPACK_PACKAGE_DESCRIPTION}\n \
> > +This is the Development package for Kodi visualization add-ons.")
>
> Uhm... are you talking about using
> CPACK_DEBIAN_component_name_PACKAGE_CONTROL_EXTRA with per component
> control file?
>
> If I understand it correctly, the control file is an all or nothing
> approach, i.e. all fields must be populated there.
>
> That brings in another step:
> configure control_component_name.in files to generate a control file per
> component with
> Version: ${PACKAGE_VERSION}, Architecture: ${CPACK_SYSTEM_NAME} and
> perhaps also Maintainer: ${PACKAGE_MAINTAINER} (or alternatively hardcode
> maintainer) during configuration.
>
> Is that what you mean?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/9987/files/c4d6cbc98e4e9ea2c4a94d9af35aa61498c73b18..11ff2d10a1fe869076b68417ab4c43f8e9f0a56f#r67348291>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFktYBVRaN38N6PLpSbGa1MXZpfywVq8ks5qMVYWgaJpZM4I3Vsx>
> .
>

wsnipex replied:
why not use the same control file as in our official packaging?

wsnipex commented:
as for generating changelog. I use this simple bash function: https://github.com/wsnipex/xbmc-packaging-ws/blob/master/buildscripts/buildfromgit.sh#L151-L172

hudokkow committed:
[cmake] Nuke legacy XBMCConfig.cmake

hudokkow commented:
Second iteration:
- separated metadata from logic as per @notspiff's suggestion
- added source tar.gz package generation
- lintian is happy (apart from warnings)
- nuked XBMCConfig.cmake

@wsnipex, any reason why nightly packages use <code>PACKAGE_PROVIDES xbmc-audio-dev</code> instead of <code>PACKAGE_PROVIDES kodi-audio-dev</code>?

hudokkow committed:
[cpack] Add Debian packages metadata

hudokkow committed:
[cmake] Build kodi-wiiremote

hudokkow committed:
[cmake] Add Debian packaging documentation

hudokkow commented:
Okido. I'm done with this one, hopefully.

Third iteration:
- added deb packaging documentation
- build missing kodi-wiiremote binary
- cleaned up...

Extensively tested during the last few days.

@fetzerch, care to do a review?

hudokkow commented:
Do not merge this.
I forgot to fetch my origin first and rebased on wrong branch. I'll fix it tomorrow when I get to the office.
Not enough ale, I tell you...

hudokkow commented:
If no objections, this is going in tonight.
jenkins cmake build: http://jenkins.kodi.tv/job/LINUX-64/8423/

fetzerch commented on project/cmake/cpack/CPackConfigDEB.cmake:
```diff
@@ -0,0 +1,350 @@
+if(CMAKE_BUILD_TYPE STREQUAL Debug)
+  set(PACKAGE_NAME_VERSION ${APP_VERSION_MAJOR}.${APP_VERSION_MINOR}~git${PACKAGE_TIMESTAMP}-${RELEASE_IDENTIFIER}-${DISTRO_CODENAME})
+else()
+  set(PACKAGE_NAME_VERSION ${APP_VERSION_MAJOR}.${APP_VERSION_MINOR}~git${PACKAGE_TIMESTAMP}-${RELEASE_IDENTIFIER}-${DISTRO_CODENAME})
```
isn't that the same as in 73? then no if/else-branch would be needed

hudokkow replied:
crap! bad rebase! will correct.
Thanks!

fetzerch commented on project/cmake/cpack/CPackConfigDEB.cmake:
```diff
@@ -0,0 +1,350 @@
+# package description common to all packages
+if(NOT CPACK_DEBIAN_PACKAGE_DESCRIPTION)
+  file(STRINGS ${PROJECT_SOURCE_DIR}/cpack/deb/package-description.txt DESC_LINES)
+  foreach(LINE ${DESC_LINES})
```
just a minor, feel free to ignore: `foreach(LINE IN LISTS DESC_LINES)`.

fetzerch commented on project/cmake/cpack/CPackConfigDEB.cmake:
```diff
@@ -0,0 +1,350 @@
+    message(FATAL_ERROR "DEB Generator: Mandatory variable CPACK_DEBIAN_${COMPONENT}_PACKAGE_PRIORITY is not set.")
+  endif()
+
+  if(DEB_PACKAGE_SHLIBDEPS STREQUAL ON)
```
the STREQUAL ON shouldn't be needed: https://cmake.org/cmake/help/latest/command/if.html

fetzerch commented on project/cmake/cpack/deb/postinst:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+set -e
+if [ "$1" = "configure" ] && [ -x "`which update-menus 2>/dev/null`" ]; then
```
`$()` instead of ``?

fetzerch commented on project/cmake/modules/FindBluetooth.cmake:
```diff
@@ -0,0 +1,29 @@
+#  BLUETOOTH_INCLUDE_DIRS - where to find bluetooth/bluetooth.h
+#  BLUETOOTH_LIBRARIES - the libraries to link against
+
+if(BLUETOOTH_LIBRARIES AND BLUETOOTH_INCLUDE_DIRS)
```
where would they come from? might be worth a comment

fetzerch commented on project/cmake/modules/FindBluetooth.cmake:
```diff
@@ -0,0 +1,29 @@
+
+if(BLUETOOTH_LIBRARIES AND BLUETOOTH_INCLUDE_DIRS)
+  set(BLUETOOTH_FOUND TRUE)
+else()
```
Ah, I guess you check them because of the variable caching:

The _DIRS and _LIBRARIES vars should not be cached (which find_* does). Cached vars won't ever be updated, so if you have to add another folder or lib to them, you have to remove CMakeCache.txt. It's better to use find_path/find_libraries with different var (convention: _INCLUDE_DIR and _LIBRARY). And also mark those as advanced then.

Any reason for not using FindPackageHandleStandardArgs? That takes care of REQUIRED, QUIETLY and sets _FOUND.

You can use https://github.com/xbmc/xbmc/blob/556636b0c9a91856d05a02c22ef4ad0836610747/project/cmake/modules/FindVDPAU.cmake as a reference for find modules. The documentation has some more details: https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#find-modules

fetzerch commented on project/cmake/scripts/linux/Install.cmake:
```diff
@@ -151,33 +323,18 @@ foreach(subdir ${build_dirs})
     string(REPLACE " " ";" subdir ${subdir})
     list(GET subdir 0 id)
-    install(CODE "execute_process(COMMAND make -C ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/${id}/src/${id}-build install DESTDIR=${DESTDIR})")
+    install(CODE "execute_process(COMMAND ${CMAKE_MAKE_PROGRAM} -C ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/${id}/src/${id}-build install DESTDIR=${DESTDIR})")
```
${CMAKE_MAKE_PROGRAM} will always point to the command line make tool for the generator. This will break ninja. 

@Fneufneu: Can you tell us why `make` doesn't work for you on BSD? Guess make exists there as well.

akva2 replied:
because 'make' on freebsd is not gnu make. and the generated makefiles need gnu make, aka 'gmake'.

fetzerch commented:
Sorry that it took so long to review. Great work! (found mostly minors, but it would be great if you could fix the find modules)

hudokkow commented:
Fixed cmake minors.
I'll have a look at the find modules / pre and post scripts tonight.

Thanks for the review!

hudokkow commented:
Minors addressed, Find modules rewritten and retested locally and with [jenkins](http://jenkins.kodi.tv/job/LINUX-64/8435/)
Going in in 3... 2...
